dm/master/scheduler/scheduler.go

```diff
@@ -384,6 +405,111 @@ func (s *Scheduler) GetSourceCfgByID(source string) *config.SourceConfig {
 	return &clone
 }
 
+// transferWorkerAndSource transfer worker and source bound relations.
+// lworker, "", "", rsource				This means a unbounded source bounded to a free worker
+// lworker, lsource, rworker, "" 		This means transfer a source from a worker to another free worker
+// lworker, lsource, "", rsource		This means transfer a worker from a bounded source to another unbounded source
+// lworker, lsource, rworker, rsource	This means transfer two bounded relations.
+func (s *Scheduler) transferWorkerAndSource(lworker, lsource, rworker, rsource string) error {
+	var (
+		lw           *Worker
+		rw           *Worker
+		boundWorkers []string
+		bounds       []ha.SourceBound
+		bound1       ha.SourceBound
+		bound2       ha.SourceBound
+	)
+
+	s.logger.Info("transfer source and worker", zap.String("left worker", lworker), zap.String("left source", lsource), zap.String("right worker", rworker), zap.String("right source", rsource))
+
+	if lworker != "" {
+		lw = s.workers[lworker]
+	}
+	if rworker != "" {
+		rw = s.workers[rworker]
+	}
+
+	// get current bounded workers.
+	if lworker != "" && lsource != "" {
+		boundWorkers = append(boundWorkers, lworker)
+	}
+	if rworker != "" && rsource != "" {
+		boundWorkers = append(boundWorkers, rworker)
+	}
+
+	// del current bounded relations.
+	if _, err := ha.DeleteSourceBound(s.etcdCli, boundWorkers...); err != nil {
+		return err
+	}
+
+	if lsource != "" {
+		s.updateStatusForUnbound(lsource)
+		s.unbounds[lsource] = struct{}{}
+	}
+	if rsource != "" {
+		s.updateStatusForUnbound(rsource)
+		s.unbounds[rsource] = struct{}{}
+	}
+
+	// put new bounded relations.
+	if lworker != "" && rsource != "" {
+		bound1 = ha.NewSourceBound(rsource, lworker)
+		bounds = append(bounds, bound1)
+	}
+	if rworker != "" && lsource != "" {
+		bound2 = ha.NewSourceBound(lsource, rworker)
+		bounds = append(bounds, bound2)
+	}
+	if _, err := ha.PutSourceBound(s.etcdCli, bounds...); err != nil {
+		return err
+	}
+
+	// update worker status
+	if lworker != "" && rsource != "" {
+		_ = s.updateStatusForBound(lw, bound1)
+		delete(s.unbounds, rsource)
+	}
+	if rworker != "" && lsource != "" {
+		_ = s.updateStatusForBound(rw, bound2)
```

Comment:
(it's quite strange that `updateStatusForBound` will update `s.bounds` inside the function, but `updateStatusForUnbound` not updating `s.unbounds`, we might fix this in another PR)

Reply:
Agree